### PR TITLE
Implement 'transactional' annotation for runtime functions.

### LINKED ISSUE #6587: Atomic transaction annotation for dispatchable calls
This issue is a syntax proposal to support atomic transaction annotation for dispatchable calls definition in `decl_module`.

### The boilerplate

With nested storage supported #6269 , now transaction rollback on error is possible. A typical atomic dispatch call implementation would be:

```rust
use frame_support::storage::{with_transaction, TransactionOutcome};

pub fn my_call(origin) {
        with_transaction(|| {
                // `my_call` implementation
                // -- snip --
                // Finally we have a `Result` of execution as `r`
		if r.is_ok() {
			TransactionOutcome::Commit(res)
		} else {
			TransactionOutcome::Rollback(res)
		}
	})?;
}
```

This pattern would be repeated every time defining an atomic dispatchable call.

### The syntax

To avoid repeating the boilerplate, I propose to add an `atomic` (or other names if preferred) annotation syntax, which will wrap the implementation in a `with_transaction` commit/rollback pattern as above. The syntax is like:

```rust
#[atomic]
pub fn my_call(origin) {
        // `my_call` implementation
        // -- snip --
}
```

And expanded as:

```rust
use frame_support::storage::{with_transaction, TransactionOutcome};

pub fn my_call(origin) {
        with_transaction(|| {
                let r = {
                        // `my_call` implementation
                        // -- snip --
                };
		if r.is_ok() {
			TransactionOutcome::Commit(res)
		} else {
			TransactionOutcome::Rollback(res)
		}
	})?;
}
```

### PR DESCRIPTION
**Pull request by shaopengw**
Provides a convenient way to make runtime functions atomic. All changes to storage performed by the annotated function are discarded if it returns `Err`, or committed if `Ok`.

Closes #6587 

TODO
- [x] Support dispatchable calls definition in `decl_module`.

### CODE REVIEW

**Comment by shaopengw** [comment]
> yeah, makes perfect sense. does it work in the macro?

Thanks for reviewing. You mean the `decl_module` macro? It would take some efforts to integrate it into `decl_module`, which has quite a few pattern matching branches. (The integration should be similar with `weight` attribute support.)

I didn't include the `decl_module` support in this PR, to make the idea easier to review and got feedback early. Would make another PR for `decl_module` support if this one accepted. :]

**Comment by shawntabrizi** [comment]
@shaopengw its just a little strange to me to merge this PR when it is not usable.

**Comment by shaopengw** [comment]
> @shaopengw its just a little strange to me to merge this PR when it is not usable.

My original thought was normal functions outside `decl_module` support is also an use case. But yeah it's only fully usable if macro support included. I'll keep working and include the macro part in this one.

**Comment by shaopengw** [comment]
Updated. It works in the `decl_module` macro now. :]

**Comment by thiolliere** [comment]
seems to be related to https://github.com/paritytech/substrate/issues/2782 also

**Comment by thiolliere** [comment]
indeed the attribute order is:
```rust
		$(#[doc = $doc_attr:tt])*
		#[weight = $weight:expr]
		$(#[$fn_attr:meta])*
```
so doc then weight then function attribute.

**Comment by thiolliere** [comment]
So at least doc of decl_module should now state that attribute on function are:
* optional doc attributes,
* then weight attributes,
* then optional function attributes, those function attributes will be written only on the dispatchable function implemented on Module, not on the Call enum variant.

At some point we could also do it more like the `try!` macro, instead of attribute macro:
```rust
let my_ok_result = try_or_rollback!(
    ensure_origin(origin)?;
    ...
);
```

**Comment by thiolliere** [comment]
if attribute order is not too magical for user, implementation is ok to me.

**Comment by shawntabrizi** [comment]
> if attribute order is not too magical for user, implementation is ok to me.

If this is okay from your perspective, it is okay to me too. I just thought it would not be expected that attribute order should matter. (Since it doesnt for normal rust)

**Comment by shaopengw** [comment]
The pattern matching nature of `macro_rules` makes the order matters. (Actually, before this PR, `doc` and `weight` also have magical order.) Maybe the procedural macro attribute refactor in #5678 could make the order arbitrary?

File: frame/support/procedural/src/transactional.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2020 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline frame/support/procedural/src/transactional.rs:3]:
```suggestion
// Copyright (C) 2020 Parity Technologies (UK) Ltd.
```

**Comment by kianenigma** [comment]
@shaopengw thanks! the build error should be irrelevant. If possible merge master one last time and we're good to go.

**Comment by shaopengw** [comment]
@kianenigma Thanks for reviewing. Updated :)

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**